SecurityExceptions thrown in Panache Before/After Listeners result in Internal Server errors

### Describe the bug

Throwing an io.quarkus.security.ForbiddenException in onBeforeAdd listener results in an Internal Server error.

### Expected behavior

The endpoint should return 403 with the message provided to the ForbiddenException 

### Actual behavior

A 500 response is sent.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

I now get a different exception:
```
Expected status code <403> but was <500>.

        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:484)
        at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:73)
        at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:108)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
        at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:512)
        at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
        at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:696)
        at io.restassured.internal.ResponseSpecificationImpl.this$2$validateResponseIfRequired(ResponseSpecificationImpl.groovy)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:198)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:62)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
        at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:135)
        at io.restassured.specification.ResponseSpecification$statusCode$0.callCurrent(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
        at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:143)
        at io.restassured.internal.ValidatableResponseOptionsImpl.statusCode(ValidatableResponseOptionsImpl.java:89)
        at org.acme.PersonResourceTest.testAdd(PersonResourceTest.java:19)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:1014)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:828)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        Suppressed: io.quarkus.rest.data.panache.RestDataPanacheException: Failed to add an entity
                at org.acme.PersonResourceJaxRs_3e3f2b08bc61477f296e187a31d4c409f4ab1ae3.add(Unknown Source)
                at java.base/java.lang.reflect.Method.invoke(Method.java:578)
                at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:154)
                at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:118)
                at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:560)
                at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:452)
                at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:413)
                at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
                at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:415)
                at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:378)
                at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:356)
                at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:70)
                at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:429)
                at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:240)
                at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:154)
                at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
                at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:157)
                at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:229)
                at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:82)
                at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:147)
                at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
                at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
                at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
                at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
                at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
                at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
                at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
                at java.base/java.lang.Thread.run(Thread.java:1589)
        Caused by: io.quarkus.security.ForbiddenException: Cannot add a person
                at org.acme.PersonResourceListener.onBeforeAdd(PersonResourceListener.java:11)
                at org.acme.PersonResourceListener.onBeforeAdd(PersonResourceListener.java:7)
                at org.acme.PersonResourceListener_ClientProxy.onBeforeAdd(Unknown Source)
                at org.acme.PersonResourceImpl_3e3f2b08bc61477f296e187a31d4c409f4ab1ae3.add(Unknown Source)
                at org.acme.PersonResourceImpl_3e3f2b08bc61477f296e187a31d4c409f4ab1ae3_Subclass.add$$superforward(Unknown Source)
                at org.acme.PersonResourceImpl_3e3f2b08bc61477f296e187a31d4c409f4ab1ae3_Subclass$$function$$2.apply(Unknown Source)
                at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
                at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
                at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)
                at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
                at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
                at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
                at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
                at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
                at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
                at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
                at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
                at org.acme.PersonResourceImpl_3e3f2b08bc61477f296e187a31d4c409f4ab1ae3_Subclass.add(Unknown Source)
                at org.acme.PersonResourceImpl_3e3f2b08bc61477f296e187a31d4c409f4ab1ae3_ClientProxy.add(Unknown Source)
                ... 28 more
```

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/ForbiddenItemListener.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/ForbiddenItemListener.java
@@ -1,0 +1,15 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.entity;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import io.quarkus.hibernate.orm.rest.data.panache.RestDataResourceMethodListener;
+import io.quarkus.security.ForbiddenException;
+
+@ApplicationScoped
+public class ForbiddenItemListener implements RestDataResourceMethodListener<AbstractItem> {
+
+    @Override
+    public void onBeforeAdd(AbstractItem item) {
+        throw new ForbiddenException("Adding items is forbidden");
+    }
+}

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/PanacheEntityResourceListenerExceptionTest.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/PanacheEntityResourceListenerExceptionTest.java
@@ -1,0 +1,29 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.entity;
+
+import static io.restassured.RestAssured.given;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusExtensionTest;
+
+class PanacheEntityResourceListenerExceptionTest {
+
+    @RegisterExtension
+    static final QuarkusExtensionTest TEST = new QuarkusExtensionTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(Collection.class, CollectionsResource.class, AbstractEntity.class,
+                            AbstractItem.class, Item.class, ItemsResource.class,
+                            ForbiddenItemListener.class)
+                    .addAsResource("application.properties")
+                    .addAsResource("import.sql"));
+
+    @Test
+    void shouldReturnForbiddenWhenListenerThrowsForbiddenException() {
+        given().accept("application/json")
+                .and().contentType("application/json")
+                .and().body("{\"name\": \"forbidden-item\", \"collection\": {\"id\": \"full\"}}")
+                .when().post("/items")
+                .then().statusCode(403);
+    }
+}

--- a/extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/entity/ForbiddenItemListener.java
+++ b/extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/entity/ForbiddenItemListener.java
@@ -1,0 +1,15 @@
+package io.quarkus.hibernate.reactive.rest.data.panache.deployment.entity;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import io.quarkus.hibernate.reactive.rest.data.panache.RestDataResourceMethodListener;
+import io.quarkus.security.ForbiddenException;
+
+@ApplicationScoped
+public class ForbiddenItemListener implements RestDataResourceMethodListener<AbstractItem> {
+
+    @Override
+    public void onBeforeAdd(AbstractItem item) {
+        throw new ForbiddenException("Adding items is forbidden");
+    }
+}

--- a/extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/entity/PanacheEntityResourceListenerExceptionTest.java
+++ b/extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/entity/PanacheEntityResourceListenerExceptionTest.java
@@ -1,0 +1,29 @@
+package io.quarkus.hibernate.reactive.rest.data.panache.deployment.entity;
+
+import static io.restassured.RestAssured.given;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusExtensionTest;
+
+class PanacheEntityResourceListenerExceptionTest {
+
+    @RegisterExtension
+    static final QuarkusExtensionTest TEST = new QuarkusExtensionTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(Collection.class, CollectionsResource.class, AbstractEntity.class,
+                            AbstractItem.class, Item.class, ItemsResource.class,
+                            ForbiddenItemListener.class)
+                    .addAsResource("application.properties")
+                    .addAsResource("import.sql"));
+
+    @Test
+    void shouldReturnForbiddenWhenListenerThrowsForbiddenException() {
+        given().accept("application/json")
+                .and().contentType("application/json")
+                .and().body("{\"name\": \"forbidden-item\", \"collection\": {\"id\": \"full\"}}")
+                .when().post("/items")
+                .then().statusCode(403);
+    }
+}

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/StandardMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/StandardMethodImplementor.java
@@ -12,6 +12,7 @@ import jakarta.ws.rs.Path;
 import jakarta.ws.rs.PathParam;
 import jakarta.ws.rs.Produces;
 import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.WebApplicationException;
 import jakarta.ws.rs.core.Context;
 
 import org.jboss.jandex.AnnotationInstance;
@@ -82,6 +83,12 @@ public abstract class StandardMethodImplementor implements MethodImplementor {
 
     protected TryBlock implementTryBlock(BytecodeCreator bytecodeCreator, String message) {
         TryBlock tryBlock = bytecodeCreator.tryBlock();
+        CatchBlockCreator waeBlock = tryBlock.addCatch(WebApplicationException.class);
+        waeBlock.throwException(waeBlock.getCaughtException());
+        waeBlock.close();
+        CatchBlockCreator secBlock = tryBlock.addCatch(SecurityException.class);
+        secBlock.throwException(secBlock.getCaughtException());
+        secBlock.close();
         CatchBlockCreator catchBlock = tryBlock.addCatch(Throwable.class);
         catchBlock.throwException(RestDataPanacheException.class, message, catchBlock.getCaughtException());
         catchBlock.close();

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/UniImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/UniImplementor.java
@@ -7,9 +7,12 @@ import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
+import jakarta.ws.rs.WebApplicationException;
+
 import org.jboss.resteasy.reactive.server.core.CurrentRequestManager;
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 
+import io.quarkus.gizmo.BranchResult;
 import io.quarkus.gizmo.BytecodeCreator;
 import io.quarkus.gizmo.FunctionCreator;
 import io.quarkus.gizmo.MethodDescriptor;
@@ -102,13 +105,30 @@ public final class UniImplementor {
 
     /**
      * Given a Uni instance, this method implements the `onFailure` method:
-     * uni.onFailure(ex -> ...).
+     * uni.onFailure().transform(ex -> ...).
+     *
+     * WebApplicationException and SecurityException are returned as-is
+     * so they can be handled by the framework's exception mappers.
+     * All other exceptions are wrapped in a RestDataPanacheException.
      */
     public static ResultHandle onFailure(BytecodeCreator creator, ResultHandle uniInstance, String message) {
-        return onFailure(creator, uniInstance,
-                (body, ex) -> body.newInstance(
-                        MethodDescriptor.ofConstructor(RestDataPanacheException.class, String.class, Throwable.class),
-                        body.load(message), ex));
+        FunctionCreator lambda = creator.createFunction(Function.class);
+        BytecodeCreator body = lambda.getBytecode();
+        ResultHandle exception = body.checkCast(body.getMethodParam(0), Throwable.class);
+
+        BranchResult isWae = body.ifTrue(body.instanceOf(exception, WebApplicationException.class));
+        isWae.trueBranch().returnValue(exception);
+        BranchResult isSec = body.ifTrue(body.instanceOf(exception, SecurityException.class));
+        isSec.trueBranch().returnValue(exception);
+
+        body.returnValue(body.newInstance(
+                MethodDescriptor.ofConstructor(RestDataPanacheException.class, String.class, Throwable.class),
+                body.load(message), exception));
+
+        ResultHandle uniOnFailure = creator.invokeInterfaceMethod(ofMethod(Uni.class, "onFailure", UniOnFailure.class),
+                uniInstance);
+        return creator.invokeVirtualMethod(ofMethod(UniOnFailure.class, "transform", Uni.class, Function.class),
+                uniOnFailure, lambda.getInstance());
     }
 
     /**

